Compare lowercase email when updating from ldap

artonge opened the pull request:
I dug into it again, and the issue is much simpler than I previously though.

- LDAP has an email address with capital letters
- NC store this address in lower case
- When the user logs in, we compare the [stored email with the new lower case email](https://github.com/nextcloud/server/blob/master/lib/private/AllConfig.php#L259-L261) before storing it. Here, both email will be the same, so we won't store the new email address with upper case letters. Which is what we want.
- We then [compare emails as they are before triggering an event](https://github.com/nextcloud/server/blob/master/lib/private/User/User.php#L202-L204), they won't match, so the user will receive an email signaling an email change every time he logs in.

The fix is to compare the old email with the new lower case email before sending the event.

~~Scenario:~~
~~- Nextcloud has an email with capital letters for a user. No idea why, as it seems we are storing it in lowercases.~~
~~- LDAP has the same email address.~~
~~- The user log in, calling [`processAttributes`](https://github.com/nextcloud/server/blob/master/apps/user_ldap/lib/User_LDAP.php#L205), which call [`updateEmail`](https://github.com/nextcloud/server/blob/master/apps/user_ldap/lib/User/User.php#L199), which [compare emails as they are stored](https://github.com/nextcloud/server/blob/af29b978078ffd9169a9bd9146feccbb7974c900/apps/user_ldap/lib/User/User.php#L443-L445) (`$currentEmail !== $email`)~~
~~- But when we are about to update the email, we once again check old and new values, but this time [we lower case the new value](https://github.com/nextcloud/server/blob/master/lib/private/AllConfig.php#L259-L261)~~

~~This means that Nextcloud will try to update the email again and again because we compare an address email with its lower case equivalent. Remember that Nextcloud has an email with capital letter in its database.~~
~~So the user will receive a continuous stream of email signaling an email change as we later [compare email as they are before triggering the event](https://github.com/nextcloud/server/blob/master/lib/private/User/User.php#L202-L204).~~

blizzz commented:
If the email already is being received lowercase it should also stay lower case. I would not know why or where this change would happen. The fix will surely work, however I am not sure this is should be necessary, when the issue might be elsewhere. Could you reproduce this behaviour?

come-nc commented on apps/user_ldap/lib/User/User.php:
```diff
@@ -440,7 +440,7 @@ public function updateEmail($valueFromLDAP = null) {
 			$user = $this->userManager->get($this->uid);
 			if (!is_null($user)) {
 				$currentEmail = (string)$user->getSystemEMailAddress();
-				if ($currentEmail !== $email) {
+				if ($currentEmail !== strtolower($email)) {
 					$user->setEMailAddress($email);
 				}
 			}
```
This seems worse somehow as the value compared against is not the one set afterwards.
If we do keep this please send strtolower($email) to the function setEMailAddress.

Also the strtolower in your change is applied to the LDAP value, not the database one, yet in your description your are talking about having uppercase in the database. This sounds inconsistent to me.

come-nc reviewed (commented):
Maybe instead fix setEMailAddress to correctly handle this usecase?
Either we support storing uppercase email and we should be able to update it. Or we do not and it should be ensured by the system.

artonge committed:
I dug into it again, and the issue is much simpler than I previously though.

- LDAP has an email address with capital letters
- NC store this address in lower case
- When the user logs in, we compare the [stored email with the new lower case email](https://github.com/nextcloud/server/blob/master/lib/private/AllConfig.php#L259-L261) before storing it. Here, both email will be the same, so we won't store the new email address with upper case letters. Which is what we want.
- We then [compare emails as they are before triggering an event](https://github.com/nextcloud/server/blob/master/lib/private/User/User.php#L202-L204), they won't match, so the user will receive an email signaling an email change every time he logs in.

The fix is to compare the old email with the new lower case email before sending the event.

Signed-off-by: Louis Chemineau <louis@chmn.me>

artonge commented:
My first analysis was wrong, I update the first comment, please review again, and thanks for the feedbacks :)

artonge commented:
/backport to stable23

artonge commented:
/backport to stable22

artonge commented:
CI failure look unrelated

artonge commented:
/rebase

artonge commented:
Just to be sure ^

timkrueger commented:
The backports for 23 and 24 are still missing because they were reverted in the past:

- https://github.com/nextcloud/server/pull/33935
- https://github.com/nextcloud/server/pull/33936

come-nc commented:
/backport to stable23

come-nc commented:
/backport to stable24
